In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

Importing Data

In [2]:
path = r'C:\Users\honoc\Documents\CareerFoundry Folder\Python Fundamentals for Data Analysts\Exercise 4.3 Folder'
path
df_prods = pd.read_csv(os.path.join(path, 'Original data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, 'Prepared data', 'orders_wrangled.csv'), index_col = False)

In [3]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Mixed-Type columns Data-Prep checks with Small Test Dataframe

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]


In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

Creating a custom function that searches the dataframe for mixed-type columns.

In [7]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_test[weird]) > 0:
        print(col)

mix


How to fix mixed-type columns. In this example, by changing other data types to string. And in the case of changing to numeric, simply update str within astype() function to int64 or whichever numeric data type you want to use.

In [8]:
df_test['mix'] = df_test['mix'].astype('str')

1. Performing Consistency Checks on df_prods Dataframe. Finding Missing Values in df_prods dataframe

In [9]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

Creating a subset of the dataframe to view these 16 missing values in the product_name column

In [10]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Addressing Missing Values by Creating a New Variable that acts like a flag based on the Missing Value.

In [11]:
df_prods.shape

(49693, 5)

Creating a New dataframe "df_prods_clean" to filter out Missing data

In [12]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False] #False because we want non-missing values in the dataframe.
df_prods_clean.shape #checking the number of rows again to see if there's a difference.


(49677, 5)

Finding Duplicates

In [13]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

Addressing Duplicates by DROPPING them after identification

Before Dropping duplicates, first check the current number of rows.

In [14]:
df_prods_clean.shape

(49677, 5)

Creating a New dataframe that doesn't include the duplicates by using the drop_duplicates() function:

In [15]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

In [16]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [17]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

The minimum value of 1 for "user_id" is reasonable, but the maximum value of 206,209,000 is extremely high, which may indicate an issue with the data, such as missing or incorrect user IDs. The minimum value of 0 and extremely high maximum value (30,000,000) for "days_since_prior_order" are both unusual and may suggest data quality issues or outliers that may require further investigation. 

3. Check for mixed-type data in your df_ords dataframe.

In [18]:
# Check for mixed types
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_ords[weird]) > 0:
        print(col)

4. No mixed-type data found

5.	Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.


In [19]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The days_since_prior_order has 206,209 missing values. This could be attributed to a number of possible reasons. One of such reasons might be First-Time Orders. Since customers who are placing their first order will not have a "days_since_prior_order" value because there is no prior order to measure the time since the last order. Secondly, Data entry errors or omissions can lead to missing values. For example, if the data recording process had issues or if certain orders were not properly recorded, it could result in missing values.

6. Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

Addressing Missing Values  in "df_ords" dataframe by Creating a New Variable that acts like a flag based on the Missing Value. But first, we view these missing values.

In [20]:
df_misval = df_ords[df_ords['days_since_prior_order'].isnull() == True]
df_misval

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

In [21]:
df_ords.shape

(3421083, 8)

In [22]:
# Creating new dataframe called df_ords_clean
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]
df_ords_clean.shape

(3214874, 8)

It is 206,209 less. I chose this method of creating an entirely new dataframe without overwriting the original version, so I can have both the treated version and the original version just in case I may still need both.

7.	Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [23]:
df_ord_dubs = df_ords_clean[df_ords_clean.duplicated()]
df_ord_dubs

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

8. No duplicates were found in "df_ords_clean" dataframe.

9.	Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [24]:
df_ords_clean

Unnamed: 0  order_id  user_id eval_set  order_number  \
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
5                 5   3367565        1    prior             6   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
1                         3                  7                    15.0  
2                         3                 12                    21.0  
3                         4                  7                    29.0  
4                         4                 15                    28.0  
5                         2                  7                    19.0  
...                     ...                ...                     ...  
3421078                   5                 18                    29.0  
3421079                   4                 10                    30.0  
3421080                   1                 12                    18.0  
3421081                   1                 12                     7.0  
3421082                   6                 14                    30.0  

[3214874 rows x 8 columns]